app/models/katello/product.rb

```diff
@@ -263,5 +264,12 @@ def ensure_provider_type_matches_id
         errors.add(:base, _("Cannot associate a Red Hat provider with a custom product"))
       end
     end
+
+    def ensure_no_sync_plans_on_empty_product
+      return if sync_plan_id.nil?
+      if repositories.empty?
+        fail _("Cannot add empty product %s to sync plan!") % name
```

Comment:
Why not `errors.add`?

I feel like if this just raises an error (`fail`) rather than adding errors, it's not a validation and shouldn't be triggered by `validate`.

Reply:
> can not be overridden given we don't have a way to catch and customize the message

You can pass a custom message to `errors.add`..

Reply:
also, if you need the message to be completely custom you can `errors.add(:base` instead of `errors.add(:sync_plan`

Reply:
So the errors.add(:base,..) makes the product record invalid and fails to save sync_plan but the error thrown is default activerecord error from here: https://apidock.com/rails/v6.0.0/ActiveRecord/Associations/CollectionAssociation/replace_records

![Screenshot from 2023-09-13 15-20-21](https://github.com/Katello/katello/assets/21146741/a36ab2af-15c5-45a7-9526-6a8871b4a4c2)

Reply:
Ah.. 

I still think `errors.add` is a more correct way to handle it.

I see it raises a `RecordNotSaved`. Maybe we could rescue that and re-raise it just to make the message better..

Reply:
Also,

Would it help to add `validates_associated :products` to `sync_plan.rb`? Seems that should have been there already. And you can add a `message:` there.